upup/models/cloudup/resources/addons/authentication.aws/k8s-1.12.yaml.template

```diff
@@ -128,9 +128,17 @@ spec:
         image: {{ or .Authentication.Aws.Image "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon/aws-iam-authenticator:v0.5.1-debian-stretch" }}
         args:
         - server
+        {{- if or (not .Authentication.Aws.BackendMode) (contains "MountedFile" .Authentication.Aws.BackendMode) }}
         - --config=/etc/aws-iam-authenticator/config.yaml
+        {{- else }}
+        - --cluster-id={{ .MasterPublicName }}
```

Comment:
good call that this flag needs to be set for the other backend modes. two things:
1. Previously the user was able to choose their cluster ID because they have full control over the configmap. If we're now hardcoding the cluster ID we should at least document it in the release notes, but even with that I think the migration from one cluster ID to another would be a bit of a pain for users: once they run `kops update cluster --yes` they'll need to have every client update their kubecfg file. Rather than dealing with this I wonder if it would be easier to just expose a ClusterID field in the API as well. If `BackendMode` is set but `ClusterID` is not set it could default to MasterPublicName. Even the [example configmap in our docs](https://kops.sigs.k8s.io/authentication/#aws-iam-authenticator) says "good choices are a random token or a domain name that will be unique to your cluster" so we shouldn't assume that existing clusters have their `clusterID == MasterPublicName`. 


2. this flag will _not_ be set if the backend mode contains MountedFile. Do we know if authenticator will recognize the cluster ID from the mounted file if `--backend-mode=CRD,MountedFile`? If it does not, we should consider setting this flag if BackendMode is set to any value, not just values that dont contain MountedFile (though then we need to look at what happens if cluster ID flag is set and `--backend-mode=MountedFile`, which cluster ID will it use?) I can take a look at the authenticator code to try to confirm the behavior there.

Reply:
So authenticator uses viper for managing configuration. It has a [--cluster-id flag as well as an env var](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/2c936a2108a739244311ac979eafaac2de142164/cmd/aws-iam-authenticator/root.go#L67), and that is combined with the [contents of the config file](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/1cfe2a90f68381eacd7b6dcfa2bf689e76eb8b4b/cmd/aws-iam-authenticator/root.go#L79) if `--config` is set. [Heres the order in which viper looks for values](https://github.com/spf13/viper/blob/e34fb51dd7c8546211df464b3f0b11bae80c8468/README.md#why-viper), so the cluster ID flag or env var will take precedence over the value in the mounted file. To reduce confusion I'm leaning towards:

1. Include `--cluster-id` flag if either `ClusterID` or `BackendMode` api fields are set (fallback to MasterPublicName if ClusterID is not set)
2. Remove the cluster ID field from the example yaml in the docs and update the docs to mention setting the api field. Also mention the fallback behavior.

This will make the change fully backwards compatible - anyone that upgrades but doesn't set their BackendMode or ClusterID api fields will continue with their existing behavior, no new flags will be set in their manifest.

Reply:
I've been doing more reading on how authenticator works. To the points in your first reply above, authenticator will read the config file regardless of what type of backend is set. I'm actually not sure any of these changes are really necessary to address the original issue 9468 since the backend is configurable in the configmap, so if someone wanted to use CRDs for the backend they could set that in their configmap and authenticator would only look for user mapping there and would ignore any user mapping that may exist in the configmap. If the backend mode is set to `CRD,MountedFile` autheticator will look for mappings in both places, but if there are conflicting entries for the same user/role/account the earlier backend in the list will take precedence.

The only way someone upgrading would run into a changing cluster-id would be if they set a backend that didn't include MountedFile and were using something other than their cluster's public API DNS address. If someone was in that position, they could work around it by including MountedFile as a backend and not having any mappings in there, but keep the other configuration. 

I agree that exposing the clusterID makes sense and setting the `--cluster-id` only when ClusterID or BackendMode is set in the kops manifest. I think we should probably also put some logic around the `--config` flag. If someone standing up a new cluster plans on using CRDs as the backend mode and the ClusterID gets set with a cli flag, there's no reason to require them to create what would be a basically empty configmap. I think the easiest way would probably be to just add a boolean `useConfigFile` or something but only evaluate it if BackendMode is set and doesn't contain MountedFile.

Reply:
That makes sense to me. Not setting `--config` would be good because it drops the requirement that the configmap exist before the authenticator pods are launched. they'll now be able to pass kops cluster validation even if no `iamidentitymapping` resources have been created yet which i think is good. I wonder if we need a boolean or if we could simply include the flag if BackendMode is set and contains MountedFile. It's less permissive than we could be but I think the simplicity will be more beneficial for a majority of the use cases. If someone wants to use CRDs but also wants to keep their `--config` configmap to use for other settings, they can just include MountedFile in their list of backend modes. We'll want to document this behavior as well.